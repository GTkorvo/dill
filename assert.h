/*
** Cthreads version of assert.h.  Using gcc version of assert.h *requires*
** the final program to be linked with GNU libc.a.  This is an unacceptable
** restriction (or at least the cause of much confusion in the use of this 
** library.  By using this version of assert.h we avoid much pain.
*/
#undef assert
#undef __assert

#ifdef NDEBUG
#define assert(ignore) ((void) 0)
#else

#define assert(expression)  \
  ((void) ((expression) ? 0 : __assert (expression, __FILE__, __LINE__)))

#ifdef LINUX_KERNEL_MODULE
#define ABORT
#else
#if !defined(_MSC_VER)
extern void abort(void);
#endif
#define ABORT abort(),
#endif
#define __assert(expression, file, lineno)  \
  (printf ("%s:%u: failed assertion\n", file, lineno),	\
   ABORT 0)

#endif
