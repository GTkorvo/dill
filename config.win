/* config.h.  Generated by configure.  */
/* config.h.in.  Generated from configure.ac by autoheader.  */

/* Define if bfd functions use bfd_byte arguments. */
/* #undef BFD_BYTE */

/* Define to 1 if the emulator should be built */
/* #undef BUILD_EMULATOR */

/* Define if dill should attempt to DCG */
/* #undef EMULATION_ONLY */

/* Define to 1 if you have the <dis-asm.h> header file. */
#define HAVE_DIS_ASM_H 1

/* Define to 1 if you have the <dlfcn.h> header file. */
#define HAVE_DLFCN_H 1

/* Define to 1 if you have the <inttypes.h> header file. */
#define HAVE_INTTYPES_H 1

/* Define to 1 if you have the <malloc.h> header file. */
#define HAVE_MALLOC_H 1

/* Define to 1 if you have the <memory.h> header file. */
#define HAVE_MEMORY_H 1

/* Define if you have the `print_insn_arm' function. */
/* #undef HAVE_PRINT_INSN_ARM */

/* Define if you have the `print_insn_i386' function. */
/* #undef HAVE_PRINT_INSN_I386 */

/* Define if you have the `print_insn_ia64' function. */
/* #undef HAVE_PRINT_INSN_IA64 */

/* Define if you have the `print_insn_little_arm' function. */
/* #undef HAVE_PRINT_INSN_LITTLE_ARM */

/* Define if you have the `print_insn_sparc' function. */
/* #undef HAVE_PRINT_INSN_SPARC */

/* Define to 1 if you have the <stdint.h> header file. */
#define HAVE_STDINT_H 1

/* Define to 1 if you have the <stdlib.h> header file. */
#define HAVE_STDLIB_H 1

/* Define to 1 if you have the <strings.h> header file. */
#define HAVE_STRINGS_H 1

/* Define to 1 if you have the <string.h> header file. */
#define HAVE_STRING_H 1

/* Define to 1 if you have the <sys/stat.h> header file. */
#define HAVE_SYS_STAT_H 1

/* Define to 1 if you have the <sys/types.h> header file. */
#define HAVE_SYS_TYPES_H 1

/* Define to 1 if you have the <unistd.h> header file. */
/* #undef HAVE_UNISTD_H */

/* Define if the host processor is ARM */
/* #undef HOST_ARM */

/* Define if the host processor is an ia64 */
/* #undef HOST_IA64 */

/* Define if the host processor is a sparc */
/* #undef HOST_SPARC */

/* Define if the host processor is a sparcv9 */
/* #undef HOST_SPARCV9 */

/* Define if the host processor is an x86 */
#define HOST_X86 1

/* Define if the host processor is an x86_64 */
/* #undef HOST_X86_64 */

/* Define if INIT_DISASSEMBLE_INFO takes three arguments instead of two */
#define INIT_DISASSEMBLE_INFO_THREE_ARG 1

/* Define if integrating with kernel plugins */
/* #undef KPLUGINS_INTEGRATION */

/* Define if compiling for linux kernel */
/* #undef LINUX_KERNEL_MODULE */

/* Define if you want more than just native target */
/* #undef MULTI_TARGET */

/* Define for the host architecture type */
#define NATIVE_ARCH "x86"

/* Define if there is no disassembler */
#define NO_DISASSEMBLER 1

/* Define if we should not use inlined procedures from BFD */
/* #undef NO_INLINED_BFD_PROCS */

/* Name of package */
#define PACKAGE "dill"

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT ""

/* Define to the full name of this package. */
#define PACKAGE_NAME ""

/* Define to the full name and version of this package. */
#define PACKAGE_STRING ""

/* Define to the one symbol short name of this package. */
#define PACKAGE_TARNAME ""

/* Define to the version of this package. */
#define PACKAGE_VERSION ""

/* The number of bytes in type long */
#define SIZEOF_LONG 4

/* Define to 1 if you have the ANSI C header files. */
#define STDC_HEADERS 1

/* Define if the membar instruction should be used to sync Icache and Dcache
   */
/* #undef USE_MEMBAR */

/* Define this is mmap should be used instead of malloc() for code memory */
/* #undef USE_MMAP_CODE_SEG */

/* Version number of package */
#define VERSION "1.0"

/* Define if byteorder is bigendian */
/* #undef WORDS_BIGENDIAN */

/* Define to `__inline__' or `__inline' if that's what the C compiler
   calls it, or to nothing if 'inline' is not supported under any name.  */
#ifndef __cplusplus
/* #undef inline */
#endif
