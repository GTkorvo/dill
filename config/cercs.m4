dnl  Mon Jan 28 13:01:16 EST 2008
dnl
dnl cercs_require_package(package, include_file, library_file)
dnl   either include file or library_file may be left off if not needed
dnl   this macro searches for the files using find_cercs_file().  If
dnl   found, it adds directory in which the found file resides to either 
dnl   CPPFLAGS (with a -I prefix)  or LDFLAGS (with a -L prefix)
dnl
AC_DEFUN([CERCS_REQUIRE_PACKAGE],
[
AC_REQUIRE([CERCS_SET_ARCHIVE])
AC_REQUIRE([CERCS_HAS_CYGPATH])
if test -z "$with_local_specified"; then
CERCS_REQUIRE_OTHER_PACKAGE($1,$2,$3)
else
CERCS_REQUIRE_OTHER_PACKAGE($1,$2,$3,1)
fi
])
dnl
dnl
AC_DEFUN([CERCS_REQUIRE_OTHER_PACKAGE],
[
AC_REQUIRE([CERCS_SET_ARCHIVE])
AC_REQUIRE([CERCS_HAS_CYGPATH])
AC_ARG_WITH(translit($1, `/',`_'), translit([  --with-$1=DIR	Where to find $1 package], `/',`_'))
define([with_translit], translit(with_$1, `/',`_'))
if test -n "$with_translit"; then
dnl
dnl if they did a with, kill the cache variables
dnl
translit(unset cercs_cv_$1_include_arg,  `/',`_')
translit(unset cercs_cv_$1_link_arg,  `/',`_')
if test `echo $with_translit | sed 's/\(.\).*/\1/g'` != "/"; then
with_translit=`pwd`/$with_translit
fi
fi
ifelse([$2], , ,
dnl
dnl  if arg2 (include_file) is specified
dnl
AC_MSG_CHECKING(needed include args for $1 package)
AC_CACHE_VAL(translit(cercs_cv_$1_include_arg, `/',`_'), 
[
ifelse([$4],1,[
search_list="$PWD/../$1 $PWD/../../$1 $PWD/../../../$1"
CERCS_SEARCH($search_list)
if test -n "$tmp_search_results"; then
cercs_tmp=$tmp_search_results
fi
],
CERCS_FIND_FILE($1, $2, cercs_tmp, $with_translit, include))
if test -n "$cercs_tmp" -a "$cercs_tmp" != "/usr/include/$2"; then
translit(cercs_cv_$1_include_arg, `/',`_')=-I`$PATHPROG $cercs_tmp | sed 's#\\\\#/#g' | sed "s#.$2##g"`
fi
])
AC_MSG_RESULT(translit($cercs_cv_$1_include_arg, `/',`_'))
dnl
dnl  add the result to CPPFLAGS if it is absent
dnl
translit(if test -n "$cercs_cv_$1_include_arg"; then, `/',`_')
translit(arg="$cercs_cv_$1_include_arg", `/',`_')
no_dash_arg=`echo $arg | sed 's/^-//g'`
[if test `echo $CPPFLAGS | grep -c "$no_dash_arg"` -eq 0; then
if test `echo $arg | grep -c "$1"` -eq 0; then
CPPFLAGS="$CPPFLAGS $arg";
else
CPPFLAGS="$arg $CPPFLAGS"
fi
fi]
fi
)
ifelse([$3], , ,
dnl
dnl  if arg3 (library_file) is specified
dnl
AC_MSG_CHECKING(needed link args for $1 package)
AC_CACHE_VAL(translit(cercs_cv_$1_link_dir,  `/',`_'), 
[
ifelse([$4],1,[
search_list="$PWD/../$1 $PWD/../../$1 $PWD/../../../$1"
CERCS_SEARCH($search_list)
if test -n "$tmp_search_results"; then
cercs_tmp=$tmp_search_results
fi
],
CERCS_FIND_FILE($1, $3, cercs_tmp, $with_translit, lib))
if test -n "$cercs_tmp" -a "$cercs_tmp" != "$3"; then
translit(cercs_cv_$1_link_dir, `/',`_')=`$PATHPROG $cercs_tmp | sed 's#\\\\#/#g' | sed "s/.$3//g"`
else
translit(cercs_cv_$1_link_dir="",  `/',`_')
fi
])
AC_MSG_RESULT(translit($cercs_cv_$1_link_dir, `/',`_'))
ld_arg="-L"
new_flags=$LDFLAGS
dnl
dnl  add the result to LDFLAGS if it is absent
dnl
translit(cercs_cv_$1_link_arg=`echo $cercs_cv_$1_link_dir, `/',`_') | sed "/./ s/^/$ld_arg/1"`
translit(if test -n "$cercs_cv_$1_link_arg"; then, `/',`_')
translit(arg=$cercs_cv_$1_link_arg, `/',`_')
no_dash_arg=`echo $arg | sed 's/^-//g'`
[if test `echo $new_flags | grep -c "$no_dash_arg"` -eq 0; then
root=`echo $3 | sed 's/\..*//'`
translit(if test ! -r $cercs_cv_$1_link_dir, `/`, `_`)/$root.la; then
if eval "test \"\${LIBTOOL+set}\" = set"; then
translit(arg="$cercs_cv_$1_link_arg "`echo $cercs_cv_$1_link_dir, `/',`_') | sed "/./ s/^/-R/1"`
fi
fi
if test `echo $arg | grep -c "$1"` -eq 0; then
# if arg does not includes a project spec add it at the end
new_flags="$new_flags $arg"
else
new_flags="$arg $new_flags"
fi
fi]
LDFLAGS=$new_flags
fi
)
])dnl
dnl
dnl cercs_find_file(package, file_to_find, variable_to_set, suggestion, dir_name)
dnl    search a set of standard directories to find file_to_find.  When found,
dnl    set $variable_to_set to the path of the file.  Use package and
dnl    suggestions to help search.  Mostly this searches ~chaos (if it exists)
dnl    and the usual suspects like:
dnl    /usr/{include,lib} /usr/local/{include,lib} /opt/<package>/{include,lib}
dnl    /opt/misc/{include,lib}.
dnl
AC_DEFUN([CERCS_FIND_FILE],
[
AC_REQUIRE([CERCS_HAS_CSH])
AC_REQUIRE([CERCS_SET_ARCHIVE])
$3=""
search_list="./$2"
CHAOS_HOMEDIR=""
if test -n "$CSH"; then
    CHAOS_HOMEDIR=`echo "echo ~chaos" | csh -sf  2>/dev/null | sed 's%/$%%'` || CHAOS_HOMEDIR=""
fi
if test -n "$4"; then
    if test `echo $4 | cut -c1` = "~"; then
        EXPANDED=`echo "echo $4" | csh -sf 2>/dev/null` || EXPANDED=""
    else
        EXPANDED=$4
    fi

    search_list="$search_list $EXPANDED/$2 $EXPANDED/$5/$2 $EXPANDED/share/$2 $EXPANDED/$cercs_cv_archive/$2 $EXPANDED/$cercs_cv_archive/$5/$2 $EXPANDED/$1/$2 $EXPANDED/$1/$cercs_cv_archive/$2 $EXPANDED/$1/$5/$2 $EXPANDED/$1/$5/$cercs_cv_archive/$2"
fi
if test -z "$with_installed_specified"; then
  search_list="$search_list `pwd`/../$1/$2 `pwd`/../$5/$2 `pwd`/../share/$2"
  if test "$CHAOS_HOMEDIR" != "$HOME"; then
    search_list="$search_list $HOME/$1/$2 $HOME/$cercs_cv_archive/$5/$2 $HOME/$5/$2"
  fi
fi
if test -n "$CHAOS_HOMEDIR" -a -n "$cercs_cv_archive"; then
  search_list="$search_list $CHAOS_HOMEDIR/$cercs_cv_archive/$1/$5/$2 $CHAOS_HOMEDIR/$cercs_cv_archive/$5/$2 $CHAOS_HOMEDIR/$1/$cercs_cv_archive/$5/$2 $CHAOS_HOMEDIR/$1/$5/$2 $CHAOS_HOMEDIR/$5/$2"
fi
if test "$libdir" != '${exec_prefix}/lib'; then
  tmpdir=`echo ${libdir} |  sed 's%/$%%'` 
  search_list="$tmpdir/$2 $search_list"
fi
if test "$exec_prefix" != "NONE"; then
  tmpdir=`echo ${exec_prefix} |  sed 's%/$%%'` 
  search_list="$tmpdir/lib/$2 $search_list"
fi
if test "$includedir" != '${prefix}/include'; then
  tmpdir=`echo ${includedir} |  sed 's%/$%%'` 
  search_list="$tmpdir/$2 $search_list"
fi
if test "$prefix" != "NONE"; then
  tmpdir=`echo ${prefix} |  sed 's%/$%%'` 
  search_list="$tmpdir/$5/$2 $search_list"
fi
if test "$5" == "lib"; then
  for tmp_lib_value in $sys_lib_search_path_spec; do
     search_list="$search_list $tmp_lib_value/$2"
  done
fi

search_list="$search_list /usr/$5/$2 /usr/local/$5/$2 /opt/$1/$5/$2 /opt/misc/$5/$2 /opt/misc/$5/$cercs_cv_archive/$2 /opt/local/$5/$2"
CERCS_SEARCH($search_list)
if test -n "$with_rpm_build_specified"; then
  if test "$5" == "lib"; then
     tmp_search_results=$libdir;
  fi
  if test "$5" == "include"; then
     tmp_search_results=$incdir;
  fi
fi
if test -n "$with_deb_build_specified"; then
# test build directories
   if test -d "../$1"; then
      tmp_search_results=`pwd`/../$1;
   fi
   if test -d "../../$1"; then
      tmp_search_results=`pwd`/../../$1;
   fi
# srcdir presence overrides
   if test -f "$srcdir/../$1/$2"; then
      tmp_search_results='${abs_top_srcdir}/../$1';
   fi
   if test -f "$srcdir/../../$1/$2"; then
      tmp_search_results='${abs_top_srcdir}/../../$1';
   fi
fi
if test -n "$tmp_search_results"; then
  $3=$tmp_search_results
fi
])dnl

AC_DEFUN([CERCS_SET_INSTALLED],[AC_ARG_WITH(installed, [  --with-installed        Don't use local copies of CERCS packages],with_installed_specified=1)])
AC_DEFUN([CERCS_SET_RPM_BUILD],[AC_ARG_WITH(rpm-build, [  --with-rpm-build        Assume necessary files in CERCS packages will be installed],with_rpm_build_specified=1)])
AC_DEFUN([CERCS_SET_DEB_BUILD],[AC_ARG_WITH(deb-build, [  --with-deb-build        Assume necessary files in CERCS packages will be installed],with_deb_build_specified=1)])
AC_DEFUN([CERCS_SET_LOCAL],[AC_ARG_WITH(local, [  --with-local            Use only local copies of CERCS packages],with_local_specified=1)])
dnl
dnl CERCS_SET_ARCHIVE()
dnl   set the $cercs_cv_machine_target variable to a standard archive name
dnl
AC_DEFUN([CERCS_SET_ARCHIVE],[
AC_REQUIRE([CERCS_SET_INSTALLED])
AC_REQUIRE([CERCS_SET_RPM_BUILD])
AC_REQUIRE([CERCS_SET_DEB_BUILD])
AC_REQUIRE([CERCS_SET_LOCAL])
AC_REQUIRE([CERCS_HAS_CSH])
CHAOS_HOMEDIR=""
if test -n "$CSH"; then
CHAOS_HOMEDIR=`echo "echo ~chaos" | csh -sf  2>/dev/null | sed 's%/$%%'` || CHAOS_HOMEDIR=""
fi
if test "$cross_compiling" = yes ; then
  cpu=$host_cpu
  vendor=$host_vendot
  os=$host_os
else
  cpu=
  vendor=
  os=
fi
if test -x $CHAOS_HOMEDIR/bin/cercs_arch; then
cercs_cv_archive=`$CHAOS_HOMEDIR/bin/cercs_arch "$cpu" "$vendor" "$os"`
else
cercs_cv_archive=${cercs_cv_archive-`cercs_arch 2>/dev/null`} || cercs_cv_archive=""
fi
])dnl
dnl
dnl  CERCS_SEARCH(variable to define, options to try)
define(CERCS_SEARCH,
[tmp_search_results=""
echo "configure:__oline__: searching for $1 " >&5
for tmp_search_value in $1; do 
   if test -r $tmp_search_value; then 
	tmp_search_results=$tmp_search_value
	echo "configure:__oline__: first found $tmp_search_results " >&5
	break
   fi 
done
])dnl
dnl
dnl
dnl CERCS_LIB_PREFIX
dnl  this macro tries to set a reasonable default for the prefix value
dnl  call with two arguments, project name and library name
dnl
AC_DEFUN([CERCS_LIB_PREFIX],
[if test "x$prefix" = xNONE; then
AC_REQUIRE([CERCS_SET_ARCHIVE])
search_list=""
CHAOS_HOMEDIR=""
if test -n "$CSH"; then
CHAOS_HOMEDIR=`echo "echo ~chaos" | csh -sf  2>/dev/null | sed 's%/$%%'` || CHAOS_HOMEDIR=""
fi
if test -n "$CHAOS_HOMEDIR"; then
search_list="$search_list $CHAOS_HOMEDIR/$cercs_cv_archive/lib/$2 $CHAOS_HOMEDIR/lib/$2"
fi
search_list="$search_list /usr/lib/$2 /usr/local/lib/$2 /opt/$1/lib/$2 /opt/misc/lib/$2"
CERCS_SEARCH($search_list)
if test -n "$tmp_search_results"; then
    prefix=`echo $tmp_search_results|sed "s%$cercs_cv_archive/lib/$2%%g;s%lib/$2%%g;s%/[^/][^/]*//*[^/][^/]*$%%"`
    exec_prefix=`echo $tmp_search_results|sed 's%lib/$2%%g;s%/[^/][^/]*//*[^/][^/]*$%%'`
fi
fi
])dnl
AC_DEFUN([CERCS_HAS_CSH], [AC_PATH_PROG(CSH,csh)])dnl
AC_DEFUN([CERCS_HAS_CYGPATH], [AC_CHECK_PROG(PATHPROG,cygpath,[cygpath -w],[echo])])dnl
