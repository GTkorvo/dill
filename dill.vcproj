<?xml version="1.0" encoding="Windows-1252"?>
<VisualStudioProject
	ProjectType="Visual C++"
	Version="8.00"
	Name="dill"
	ProjectGUID="{A525C378-C99B-41A3-861B-AD5DEA11F6EC}"
	RootNamespace="dill"
	Keyword="Win32Proj"
	>
	<Platforms>
		<Platform
			Name="Win32"
		/>
	</Platforms>
	<ToolFiles>
	</ToolFiles>
	<Configurations>
		<Configuration
			Name="Debug|Win32"
			OutputDirectory="..\bin"
			IntermediateDirectory="$(OutDir)\$(ConfigurationName)\$(ProjectName)"
			ConfigurationType="4"
			CharacterSet="1"
			>
			<Tool
				Name="VCPreBuildEventTool"
				CommandLine="prebuild-win32.bat"
			/>
			<Tool
				Name="VCCustomBuildTool"
			/>
			<Tool
				Name="VCXMLDataGeneratorTool"
			/>
			<Tool
				Name="VCWebServiceProxyGeneratorTool"
			/>
			<Tool
				Name="VCMIDLTool"
			/>
			<Tool
				Name="VCCLCompilerTool"
				Optimization="0"
				PreprocessorDefinitions="WIN32;_DEBUG;_LIB;_CONSOLE;DILL_EXPORTS;_CRT_SECURE_NO_DEPRECATE"
				MinimalRebuild="true"
				BasicRuntimeChecks="3"
				RuntimeLibrary="3"
				UsePrecompiledHeader="0"
				WarningLevel="3"
				Detect64BitPortabilityProblems="true"
				DebugInformationFormat="4"
			/>
			<Tool
				Name="VCManagedResourceCompilerTool"
			/>
			<Tool
				Name="VCResourceCompilerTool"
			/>
			<Tool
				Name="VCPreLinkEventTool"
			/>
			<Tool
				Name="VCLibrarianTool"
				OutputFile="$(OutDir)\$(ProjectName)_d.lib"
			/>
			<Tool
				Name="VCALinkTool"
			/>
			<Tool
				Name="VCXDCMakeTool"
			/>
			<Tool
				Name="VCBscMakeTool"
			/>
			<Tool
				Name="VCFxCopTool"
			/>
			<Tool
				Name="VCPostBuildEventTool"
			/>
		</Configuration>
		<Configuration
			Name="Release|Win32"
			OutputDirectory="..\bin"
			IntermediateDirectory="$(OutDir)\$(ConfigurationName)\$(ProjectName)"
			ConfigurationType="4"
			CharacterSet="1"
			WholeProgramOptimization="1"
			>
			<Tool
				Name="VCPreBuildEventTool"
				CommandLine="prebuild-win32.bat"
			/>
			<Tool
				Name="VCCustomBuildTool"
			/>
			<Tool
				Name="VCXMLDataGeneratorTool"
			/>
			<Tool
				Name="VCWebServiceProxyGeneratorTool"
			/>
			<Tool
				Name="VCMIDLTool"
			/>
			<Tool
				Name="VCCLCompilerTool"
				PreprocessorDefinitions="WIN32;NDEBUG;_LIB;_CONSOLE;DILL_EXPORTS;_CRT_SECURE_NO_DEPRECATE"
				RuntimeLibrary="2"
				UsePrecompiledHeader="0"
				WarningLevel="3"
				Detect64BitPortabilityProblems="true"
				DebugInformationFormat="3"
			/>
			<Tool
				Name="VCManagedResourceCompilerTool"
			/>
			<Tool
				Name="VCResourceCompilerTool"
			/>
			<Tool
				Name="VCPreLinkEventTool"
			/>
			<Tool
				Name="VCLibrarianTool"
			/>
			<Tool
				Name="VCALinkTool"
			/>
			<Tool
				Name="VCXDCMakeTool"
			/>
			<Tool
				Name="VCBscMakeTool"
			/>
			<Tool
				Name="VCFxCopTool"
			/>
			<Tool
				Name="VCPostBuildEventTool"
			/>
		</Configuration>
	</Configurations>
	<References>
	</References>
	<Files>
		<Filter
			Name="Source Files"
			Filter="cpp;c;cc;cxx;def;odl;idl;hpj;bat;asm;asmx"
			UniqueIdentifier="{4FC737F1-C7A5-4376-A066-2A32D752A2FF}"
			>
			<File
				RelativePath=".\dill_util.c"
				>
			</File>
			<File
				RelativePath=".\version.c"
				>
			</File>
			<File
				RelativePath=".\virtual.c"
				>
			</File>
			<File
				RelativePath=".\x86.c"
				>
			</File>
		</Filter>
		<Filter
			Name="Header Files"
			Filter="h;hpp;hxx;hm;inl;inc;xsd"
			UniqueIdentifier="{93995380-89BD-4b04-88EB-625FBE52EBFB}"
			>
			<File
				RelativePath=".\assert.h"
				>
			</File>
			<File
				RelativePath=".\dill_internal.h"
				>
			</File>
			<File
				RelativePath=".\virtual.h"
				>
			</File>
			<File
				RelativePath=".\x86.h"
				>
			</File>
		</Filter>
	</Files>
	<Globals>
	</Globals>
</VisualStudioProject>
