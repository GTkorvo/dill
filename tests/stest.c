/* Test that save/restore and locals do not overlap in the activation record. */
#include <assert.h>
#include <stdio.h>
#include <malloc.h>
#include <unistd.h>

#include "../config.h"
#include "dill.h"
#ifdef USE_MMAP_CODE_SEG
#include "sys/mman.h"
#endif

void * mk_test(dill_stream c) {
	int i, regs, fregs;
	int l[100];
	int abortl1;
	int abortl2;
	int abortl3;

	dill_reg r[32];
	dill_reg fr[32];

	dill_start_proc(c, "foo", DILL_I, "%i%d");

	abortl1 = dill_alloc_label(c);
	abortl2 = dill_alloc_label(c);
	abortl3 = dill_alloc_label(c);

	for(i = 0; i < 5 && dill_raw_getreg(c, &r[i], DILL_I, DILL_TEMP); i++) {
/*		(void)("allocated register %d\n", i);*/
		dill_seti(c, r[i], i);
		dill_savei(c, r[i]); 
	}
	regs = i;

	for(i = 0; dill_raw_getreg(c, &fr[i], DILL_D, DILL_TEMP); i++) {
/*		(void)("allocated register %d\n", i);*/
		dill_setd(c, fr[i], (double)i);
		dill_saved(c, fr[i]); 
	}
	fregs = i;

	for(i = 0; i < 10; i++) {
		l[i] = dill_local(c, DILL_I);
			
		dill_seti(c, dill_param_reg(c,0), 100 + i);
		dill_stii(c, dill_param_reg(c,0), dill_lp(c), l[i]);
	}

	/* call a procedure, then verify that everything is in place. */
	dill_scallv(c, (void*)printf, "%P", "hello world!\n");


	for(i = 0; i < regs; i++) {
		dill_restorei(c, r[i]); 
		dill_bneii(c, r[i], i, abortl1);
	}

	for(i = 0; i < fregs; i++) {
/*		(void)("allocated register %d\n", i);*/
		dill_restored(c, fr[i]); 
		dill_setd(c, dill_param_reg(c,1), (double)i);
		dill_bned(c, fr[i], dill_param_reg(c,1), abortl2);
	}

	for(i = 0; i < 10; i++) {
		dill_ldii(c, dill_param_reg(c, 0), dill_lp(c), l[i]);
		dill_bneii(c, dill_param_reg(c, 0), 100 + i, abortl3);
	}
	dill_retii(c, 0);

	dill_mark_label(c, abortl1);
	dill_retii(c, 1);		/* failure. */
	dill_mark_label(c, abortl2);
	dill_retii(c, 2);		/* failure. */
	dill_mark_label(c, abortl3);
	dill_retii(c, 3);		/* failure. */

	return dill_end(c);
}

int main() { 
	int (*ip)();
	dill_stream c = dill_create_raw_stream();
	int ret;
	char *target;

	ip = (int (*)()) mk_test(c);
	ret = ip();

	if(ret == 0) {
		printf("success!\n");
	} else {
		dill_dump(c);
		printf("failure at point %d!\n", ret);
	}
	ip = (int (*)()) mk_test(c);
	ret = ip();
	if(ret == 0)
		printf("success!\n");
	else {
		dill_dump(c);
		printf("failure at point %d (second)\n", ret);
	}
#ifdef USE_MMAP_CODE_SEG
	{
	    int size = dill_code_size(c);
	    static unsigned long ps = -1;
	    if (ps == -1) {
	        ps = (getpagesize ());
	    }
	    if (ps > size) size = ps;
	    target = (void*)mmap(0, size, 
				 PROT_EXEC | PROT_READ | PROT_WRITE, 
				 MAP_ANONYMOUS|MAP_PRIVATE, -1, 0);
	}
	if (target == (void*)-1) perror("mmap");
#else
	target = (void*)malloc(dill_code_size(c));
#endif
	ip = (int (*)()) dill_clone_code(c, target, dill_code_size(c));
	dill_free_context(c);
	ret = ip();
	if(ret == 0)
		printf("success!\n");
	else {
		dill_dump(c);
		printf("failure at point %d (third)\n", ret);
	}

	return 0;
}
