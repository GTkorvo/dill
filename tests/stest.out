hello world!
success!
hello world!
success!
hello world!
success!
